keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,708 @@
+Therefore, it won't affect the protection.
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+- Garbage collector continues to try to delete a dependent resource that lien is added
+  - Detection:
+    - Deletion of an owner of a resource isn't completed.
+    - A dependent resource isn't deleted.
+  - Mitigations: Delete the lien on the dependent resource.
+  - Diagnostics:
+    - Log messages: [message](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/garbagecollector/garbagecollector.go#L475) like below repeatedly logged.
+	  `I0908 22:00:47.800913 4063637 garbagecollector.go:475] "Processing object" object="default/nginx-deployment-66b6c48dd5" objectUID=d38e5bc2-1b10-4f08-8e54-6c6c9afbfe3c kind="ReplicaSet" virtual=false`
+	- Log Level: 2
+  - Testing: E2E test `Verify that foreground owner deletion isn't complete while dependent with blockOwnerDeletion=true and lien exists` covers this failure mode.
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+- Each controller for protecting objects implements its own logic. However, it requires much implementations for the same logic and potentially provides inconsistent user interfaces, like many different finalizers and different ways to opt-out, to users.
+- Implement similar mechanism by using finalizers or admission webhook to block deletion
```

Comment:
What are the tradeoffs between Liens and admission webhooks?  

It seems like there are many more examples of admission webhooks than there are examples of finalizer use?  Should we prefer a pattern which is more widely used?

Reply:
Admission webhooks are not very comparable in this case. Yes, you could make one that did the trick, but the real solution to this problem is a Schelling point. That is, the problem isn't "keep my object from being deleted" it's "let parties who know nothing of each other communicate the need to preserve an object." In that sense, whatever is merged upstream will win as long as it's not grossly ill-suited. And it is easier for us to add some apiserver logic than to operate a webhook.